firefox-ios/Client/Frontend/TabContentsScripts/NightModeHelper.swift

```diff
@@ -79,4 +90,44 @@ class NightModeHelper: TabContentScript, FeatureFlaggable {
     ) {
         userDefaults.removeObject(forKey: NightModeKeys.DarkThemeEnabled)
     }
+
+    static func isPageInDarkMode(webView: WKWebView?, completion: @escaping (Bool) -> Void) {
+        let timerId = GleanMetrics.DarkReader.websiteDarkThemeDetection.start()
+        takePageSnapshot(for: webView, at: 0.6, width: 300, height: 400) { color in
+            let isPageDark = color?.isDark ?? false
+            GleanMetrics.DarkReader.websiteDarkThemeDetection.stopAndAccumulate(timerId)
+            completion(isPageDark)
+        }
+    }
+
+    private static func takePageSnapshot(
+        for webView: WKWebView?,
+        at verticalOffset: CGFloat,
+        width: CGFloat,
+        height: CGFloat,
+        completion: @escaping (UIColor?) -> Void
+    ) {
+        guard let webView = webView else { return }
+
+        // Ensure width does not exceed the webView’s frame width
+        let snapshotWidth = min(width, webView.frame.width)
+        // Center horizontally
+        let xPosition = (webView.frame.width - snapshotWidth) / 2
+
+        let contentHeight = webView.scrollView.contentSize.height
+        let snapshotHeight = min(height, contentHeight)
+
+        var yPosition = contentHeight * verticalOffset
+        // Shift up if the snapshot overflows the page
+        if yPosition + snapshotHeight > contentHeight {
+            yPosition = contentHeight - snapshotHeight
+        }
+
+        let config = WKSnapshotConfiguration()
+        config.rect = CGRect(x: xPosition, y: yPosition, width: snapshotWidth, height: snapshotHeight)
+
+        webView.takeSnapshot(with: config) { (image: UIImage?, _) in
+            completion(image?.averageColor())
```

Comment:
thanks for adding the average color calculate @issammani ! Seems like we can do something similar and can take a snapshot of the top row or can poll the top pixels. Taking a snapshot might not apply in some situations, but seems like it can get us there in most cases, I'll  branch off this and let you know what I come up with!

Reply:
@issammani there is a delay in retrieving the color 🤔 , similar to the performance you mentioned in the PR description; I'm wondering if we should take a snapshot or retrieve the data earlier or if for now it's okay to show a default color in the meantime

Reply:
@cyndichin @issammani are you two still looking into this PR?

Reply:
Yes. @cyndichin and I have a call tomorrow to discuss this 😄